# HG changeset patch
# User Rik <rik@octave.org>
# Date 1533836334 25200
#      Thu Aug 09 10:38:54 2018 -0700
# Node ID 134939b92d866564cb661e05164d64f5506c7dca
# Parent  ddf8a03beffb037d7fc2a0958c245b54d8171d47
strmatch.m: Make m-file a legacy function.


* NEWS: Announce definition of new class of functions (legacy) and present a
list of legacy functions and their replacements.

* scripts/legacy/module.mk, scripts/strings/module.mk: Move strmatch.m from
strings/ dir to legacy/ dir.

* strmatch.m: Added initial warning BIST test to shut off legacy warning when
running tests.

diff --git a/NEWS b/NEWS
--- a/NEWS
+++ b/NEWS
@@ -62,16 +62,30 @@ Summary of important user-visible change
  ** The OSMesa library is no longer used.  To print invisible figures
     when using OpenGL graphics, the Qt QOFFSCREENSURFACE feature must be
     available and you must use the qt graphics toolkit.
 
  ** New functions added in 5.0:
 
       ordeig
 
+ ** Legacy functions.
+
+    The following functions have been declared legacy functions which
+    means they are obsolete and should not be used in any new code.
+    Unlike deprecated functions, however, their removal from Octave has
+    not yet been scheduled.
+
+      Function             | Replacement
+      ---------------------|------------------
+      isequalwithequalnans | isequaln
+      isstr                | ischar
+      setstr               | char
+      strmatch             | strncmp or strcmp
+
  ** Deprecated functions.
 
     The following functions have been deprecated in Octave 5.0 and will
     be removed from Octave 7 (or whatever version is the second major
     release after 5.0):
 
       Function             | Replacement
       ---------------------|------------------
diff --git a/scripts/legacy/module.mk b/scripts/legacy/module.mk
--- a/scripts/legacy/module.mk
+++ b/scripts/legacy/module.mk
@@ -1,14 +1,15 @@
 FCN_FILE_DIRS += scripts/legacy
 
 %canon_reldir%_FCN_FILES = \
   %reldir%/isequalwithequalnans.m \
   %reldir%/isstr.m \
-  %reldir%/setstr.m
+  %reldir%/setstr.m \
+  %reldir%/strmatch.m
 
 %canon_reldir%dir = $(fcnfiledir)/legacy
 
 %canon_reldir%_DATA = $(%canon_reldir%_FCN_FILES)
 
 FCN_FILES += $(%canon_reldir%_FCN_FILES)
 
 PKG_ADD_FILES += %reldir%/PKG_ADD
diff --git a/scripts/strings/strmatch.m b/scripts/legacy/strmatch.m
rename from scripts/strings/strmatch.m
rename to scripts/legacy/strmatch.m
--- a/scripts/strings/strmatch.m
+++ b/scripts/legacy/strmatch.m
@@ -16,16 +16,20 @@
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <https://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
 ## @deftypefn  {} {} strmatch (@var{s}, @var{A})
 ## @deftypefnx {} {} strmatch (@var{s}, @var{A}, "exact")
+##
+## This function is obsolete.  Use an alternative such as @code{strncmp}
+## or @code{strcmp} instead.
+##
 ## Return indices of entries of @var{A} which begin with the string @var{s}.
 ##
 ## The second argument @var{A} must be a string, character matrix, or a cell
 ## array of strings.
 ##
 ## If the third argument @qcode{"exact"} is not given, then @var{s} only
 ## needs to match @var{A} up to the length of @var{s}.  Trailing spaces and
 ## nulls in @var{s} and @var{A} are ignored when matching.
@@ -40,27 +44,35 @@
 ## strmatch ("apple", ["apple  "; "apple juice"; "an apple"])
 ##      @result{} [1; 2]
 ##
 ## strmatch ("apple", ["apple  "; "apple juice"; "an apple"], "exact")
 ##      @result{} [1]
 ## @end group
 ## @end example
 ##
-## @strong{Caution:} @code{strmatch} is scheduled for deprecation.  Use
-## @code{strncmp} (normal case), or @code{strcmp} (@qcode{"exact"} case), or
-## @code{regexp} in all new code.
-## @seealso{strfind, findstr, strcmp, strncmp, strcmpi, strncmpi, find}
+## @strong{Caution:} @code{strmatch} is obsolete.  Use @code{strncmp} (normal
+## case) or @code{strcmp} (@qcode{"exact"} case) in all new code.  Other
+## replacement possibilities, depending on application, include @code{regexp}
+## or @code{validatestring}.
+## @seealso{strncmp, strcmp, regexp, strfind, validatestring}
 ## @end deftypefn
 
 ## Author: Paul Kienzle, Alois Schloegl
 ## Adapted-by: jwe
 
 function idx = strmatch (s, A, exact)
 
+  persistent warned = false;
+  if (! warned)
+    warned = true;
+    warning ("Octave:legacy-function",
+             "strmatch is obsolete; use strncmp or strcmp instead\n");
+  endif
+
   if (nargin < 2 || nargin > 3)
     print_usage ();
   endif
 
   if (iscellstr (s))
     if (numel (s) > 1)
       error ("strmatch: a cell array S must contain only one string");
     endif
@@ -102,16 +114,21 @@ function idx = strmatch (s, A, exact)
       B = regexprep (A(idx), "[ \\0]+$", '');
       idx = idx(strcmp (s, B));
     endif
   endif
 
 endfunction
 
 
+## First test is necessary to provoke 1-time legacy warning
+%!test
+%! warning ("off", "Octave:legacy-function", "local");
+%! strmatch ("", "");
+
 %!assert (strmatch ("a", {"aaa", "bab", "bbb"}), 1)
 %!assert (strmatch ("apple", "apple juice"), 1)
 %!assert (strmatch ("apple", ["apple pie"; "apple juice"; "an apple"]), [1; 2])
 %!assert (strmatch ("apple", {"apple pie"; "apple juice"; "tomato"}), [1; 2])
 %!assert (strmatch ("apple pie", "apple"), [])
 %!assert (strmatch ("a ", "a"), 1)
 %!assert (strmatch ("a", "a \0", "exact"), 1)
 %!assert (strmatch ("a b", {"a b", "a c", "c d"}), 1)
diff --git a/scripts/strings/module.mk b/scripts/strings/module.mk
--- a/scripts/strings/module.mk
+++ b/scripts/strings/module.mk
@@ -21,17 +21,16 @@ FCN_FILE_DIRS += scripts/strings
   %reldir%/ostrsplit.m \
   %reldir%/regexptranslate.m \
   %reldir%/rindex.m \
   %reldir%/str2num.m \
   %reldir%/strcat.m \
   %reldir%/strchr.m \
   %reldir%/strjoin.m \
   %reldir%/strjust.m \
-  %reldir%/strmatch.m \
   %reldir%/strsplit.m \
   %reldir%/strtok.m \
   %reldir%/strtrim.m \
   %reldir%/strtrunc.m \
   %reldir%/substr.m \
   %reldir%/unicode2native.m \
   %reldir%/untabify.m \
   %reldir%/validatestring.m
